File: backend/src/main/java/com/skep/equipment/Equipment.java
package com.skep.equipment;

import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

@Entity
@Table(name = "equipment")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Equipment {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "supplier_id", nullable = false)
    private Long supplierId;

    @Column(name = "vehicle_no", length = 32)
    private String vehicleNo;

    @Column(nullable = false, length = 32)
    private String category;

    @Column(length = 100)
    private String model;

    @Column(length = 100)
    private String manufacturer;

    private Integer year;

    @Column(name = "photo_key", length = 255)
    private String photoKey;

    @Column(name = "photo_content_type", length = 100)
    private String photoContentType;

    @Column(length = 64)
    private String code;

    /** NFC(RFC) 차량 태그 — 카드 도착 후 등록. 일상점검/출근 시 태그로 식별. */
    @Column(name = "nfc_tag_id", length = 64)
    private String nfcTagId;

    @Column(name = "serial_number", length = 128)
    private String serialNumber;

    @Column(name = "usage_hours")
    private Integer usageHours;

    @Column(name = "weight_kg")
    private Integer weightKg;

    @Column(name = "bucket_capacity", precision = 8, scale = 2)
    private java.math.BigDecimal bucketCapacity;

    @Column(name = "insurance_expiry")
    private java.time.LocalDate insuranceExpiry;

    // V70: 차량관리 만료/교체 due (정기검사/오일/등록)
    @Column(name = "inspection_due_date")
    private java.time.LocalDate inspectionDueDate;
    @Column(name = "oil_change_due_date")
    private java.time.LocalDate oilChangeDueDate;
    @Column(name = "registration_expiry")
    private java.time.LocalDate registrationExpiry;

    // V27: 자동차등록증상 차주(소유자) — 사업자(supplier_id) 와 다를 수 있음
    @Column(name = "vehicle_owner_name", length = 100)
    private String vehicleOwnerName;
    @Column(name = "vehicle_owner_business_no", length = 32)
    private String vehicleOwnerBusinessNo;
    @Column(name = "vehicle_owner_resident_no", length = 32)
    private String vehicleOwnerResidentNo;

    // 외부 조달 여부 — 우리 공급사 장비(false) / 외부에서 가져온 장비(true). 외부면 소유자(사업자) 서류 별도.
    @Column(name = "is_external", nullable = false)
    private boolean external;

    // Phase4: 외부 장비 기사(조종원) Person 연결 — 기사 로그인 계정.
    @Column(name = "operator_person_id")
    private Long operatorPersonId;

    @Column(name = "operating_hours", nullable = false)
    private int operatingHours;

    @Column(name = "idle_hours", nullable = false)
    private int idleHours;

    @Column(name = "downtime_hours", nullable = false)
    private int downtimeHours;

    // V11: 현재 배치 정보
    @Column(name = "current_site_id")
    private Long currentSiteId;

    @Enumerated(EnumType.STRING)
    @Column(name = "assignment_status", nullable = false, length = 32)
    private EquipmentAssignmentStatus assignmentStatus = EquipmentAssignmentStatus.AVAILABLE;

    @Column(name = "last_assigned_at")
    private LocalDateTime lastAssignedAt;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    @Builder
    private Equipment(Long supplierId, String vehicleNo, String category,
                      String model, String manufacturer, Integer year) {
        this.supplierId = supplierId;
        this.vehicleNo = vehicleNo;
        this.category = category;
        this.model = model;
        this.manufacturer = manufacturer;
        this.year = year;
    }

    @PrePersist
    void onCreate() {
        this.createdAt = LocalDateTime.now();
        this.updatedAt = this.createdAt;
    }

    @PreUpdate
    void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }

    public void update(String vehicleNo, String category, String model, String manufacturer, Integer year) {
        if (vehicleNo != null) this.vehicleNo = vehicleNo;
        if (category != null) this.category = category;
        if (model != null) this.model = model;
        if (manufacturer != null) this.manufacturer = manufacturer;
        if (year != null) this.year = year;
    }

    /** 외부 조달 여부 + 소유자(차주/사업자) 정보 갱신. */
    public void updateSourcing(Boolean external, String ownerName, String ownerBusinessNo) {
        if (external != null) this.external = external;
        if (ownerName != null) this.vehicleOwnerName = ownerName.isBlank() ? null : ownerName.trim();
        if (ownerBusinessNo != null) this.vehicleOwnerBusinessNo = ownerBusinessNo.isBlank() ? null : ownerBusinessNo.trim();
    }

    /** Phase4: 외부 장비 기사(조종원) Person 연결/해제. */
    public void linkOperator(Long personId) {
        this.operatorPersonId = personId;
    }

    public void setPhoto(String key, String contentType) {
        this.photoKey = key;
        this.photoContentType = contentType;
    }

    public void clearPhoto() {
        this.photoKey = null;
        this.photoContentType = null;
    }

    /** 현장에 배치한다. assignment_status를 ASSIGNED로, current_site_id/last_assigned_at 업데이트. */
    public void assignToSite(Long siteId, LocalDateTime when) {
        this.currentSiteId = siteId;
        this.assignmentStatus = EquipmentAssignmentStatus.ASSIGNED;
        this.lastAssignedAt = when;
    }

    /** 현장 해제. assignment_status를 AVAILABLE로 되돌리되, BROKEN 상태였으면 유지. */
    public void releaseFromSite() {
        this.currentSiteId = null;
        if (this.assignmentStatus != EquipmentAssignmentStatus.BROKEN) {
            this.assignmentStatus = EquipmentAssignmentStatus.AVAILABLE;
        }
    }

    public void setAssignmentStatus(EquipmentAssignmentStatus status) {
        this.assignmentStatus = status;
    }

    public void assignNfcTag(String tagId) {
        this.nfcTagId = (tagId == null || tagId.isBlank()) ? null : tagId.trim();
    }

    public void setDueDates(java.time.LocalDate inspectionDue, java.time.LocalDate oilDue,
                            java.time.LocalDate regExpiry) {
        this.inspectionDueDate = inspectionDue;
        this.oilChangeDueDate = oilDue;
        this.registrationExpiry = regExpiry;
    }

    /** 검사만료일(정기검사 유효기간) 단건 설정 — 장비 등록 시 폼/OCR 값 저장용(다른 due 는 보존). */
    public void setInspectionDueDate(java.time.LocalDate d) {
        this.inspectionDueDate = d;
    }
}

File: frontend/src/types/equipment.ts
export type EquipmentCategory =
  | 'EXCAVATOR'
  | 'WHEEL_LOADER'
  | 'CRANE'
  | 'FORKLIFT'
  | 'DOZER'
  | 'GRADER'
  | 'AERIAL_LIFT'
  | 'PUMP_TRUCK'
  | 'ATTACHMENT'
  | 'ROLLER'
  | 'PILE_DRIVER'
  | 'CONCRETE_MIXER_TRUCK'
  | 'ASPHALT_FINISHER'
  | 'CRUSHER'
  | 'AIR_COMPRESSOR'
  | 'BORING_MACHINE'
  | 'TOWER_CRANE'
  | 'DUMP_TRUCK'
  | 'SKID_LOADER'
  | 'TRAILER'
  | 'WING_BODY'
  | 'CARGO_TRUCK'
  | 'FREIGHT_TRUCK'
  | 'TANK_LORRY'
  | 'LADDER_TRUCK';

import type { EquipmentAssignmentStatus } from './assignment';

export type EquipmentResponse = {
  id: number;
  supplier_id: number;
  supplier_name?: string | null;
  vehicle_no?: string | null;
  category: EquipmentCategory;
  model?: string | null;
  manufacturer?: string | null;
  year?: number | null;
  is_external?: boolean;
  vehicle_owner_name?: string | null;
  vehicle_owner_business_no?: string | null;
  operator_person_id?: number | null;
  has_photo: boolean;
  expiring_count: number;
  // 신규 (V8)
  code?: string | null;
  serial_number?: string | null;
  usage_hours?: number | null;
  weight_kg?: number | null;
  bucket_capacity?: number | string | null;
  insurance_expiry?: string | null;
  // V70 차량관리 due
  inspection_due_date?: string | null;
  oil_change_due_date?: string | null;
  registration_expiry?: string | null;
  operating_hours: number;
  idle_hours: number;
  downtime_hours: number;
  utilization_pct?: number | null;
  // S4'(P3a) 가동시간 기반 정비 — 누적 가동시간 / 현재 현장 정비주기(null=비활성) / 도래 여부.
  cumulative_work_hours: number;
  maintenance_interval_hours?: number | null;
  maintenance_due: boolean;
  // V11 배치 정보
  current_site_id?: number | null;
  current_site_name?: string | null;
  assignment_status?: EquipmentAssignmentStatus;
  last_assigned_at?: string | null;
  // ---
  created_at: string;
  updated_at?: string | null;
};

export const EQUIPMENT_CATEGORY_LABEL: Record<EquipmentCategory, string> = {
  EXCAVATOR: '굴삭기',
  WHEEL_LOADER: '휠로더',
  CRANE: '크레인',
  FORKLIFT: '지게차',
  DOZER: '도저',
  GRADER: '그레이더',
  AERIAL_LIFT: '고소작업차',
  PUMP_TRUCK: '펌프카',
  ATTACHMENT: '어태치먼트',
  ROLLER: '롤러',
  PILE_DRIVER: '항타·항발기',
  CONCRETE_MIXER_TRUCK: '콘크리트믹서트럭',
  ASPHALT_FINISHER: '아스팔트피니셔',
  CRUSHER: '쇄석기',
  AIR_COMPRESSOR: '공기압축기',
  BORING_MACHINE: '천공기',
  TOWER_CRANE: '타워크레인',
  DUMP_TRUCK: '덤프트럭',
  SKID_LOADER: '스키드로더',
  TRAILER: '트레일러',
  WING_BODY: '윙바디',
  CARGO_TRUCK: '카고트럭',
  FREIGHT_TRUCK: '화물차',
  TANK_LORRY: '탱크로리',
  LADDER_TRUCK: '사다리차',
};

/**
 * 종류 code → 표시 라벨. 어드민이 추가한 새 코드는 정적 맵에 없으므로 code 자체를 폴백으로 반환.
 * 새 코드의 한글 이름은 useEquipmentTypes() 훅(런타임 /api/equipment-types)으로 해석.
 */
export function equipmentCategoryLabel(code: string | null | undefined): string {
  if (!code) return '';
  return EQUIPMENT_CATEGORY_LABEL[code as EquipmentCategory] ?? code;
}

// 표시 순서: 건설기계(등록증) → 차량(자동차등록증) → 기타.
export const EQUIPMENT_CATEGORIES: EquipmentCategory[] = [
  'EXCAVATOR', 'WHEEL_LOADER', 'CRANE', 'FORKLIFT', 'DOZER', 'GRADER', 'PUMP_TRUCK',
  'ROLLER', 'PILE_DRIVER', 'CONCRETE_MIXER_TRUCK', 'ASPHALT_FINISHER', 'CRUSHER',
  'AIR_COMPRESSOR', 'BORING_MACHINE', 'TOWER_CRANE', 'DUMP_TRUCK', 'SKID_LOADER',
  'AERIAL_LIFT', 'TRAILER', 'WING_BODY', 'CARGO_TRUCK', 'FREIGHT_TRUCK', 'TANK_LORRY', 'LADDER_TRUCK',
  'ATTACHMENT',
];
